test/src/test/java/hudson/bugs/JnlpAccessWithSecuredHudsonTest.java

```diff
@@ -95,8 +95,8 @@ public void anonymousCanAlwaysLoadJARs() throws Exception {
         XmlPage jnlp = (XmlPage) wc.goTo("computer/test/slave-agent.jnlp","application/x-java-jnlp-file");
         URL baseUrl = jnlp.getUrl();
         Document dom = new DOMReader().read(jnlp.getXmlDocument());
-        for( Element jar : (List<Element>)dom.selectNodes("//jar") ) {
-            URL url = new URL(baseUrl,jar.attributeValue("href"));
+        for( Object jar : dom.selectNodes("//jar") ) {
```

Comment:
@jeffret-b is this because the return type of the selectNodes() method has changed in the new implementation?

Reply:
Basically.

They started using generics between the ancient dom4j version 1.6 and current. Before it just returned `List` and because of the backwards compatibility around generics in Java, the compiler was happy to cast it to a list of `Element`. But, it really returns a `List<Node>` and the compiler isn't happy about trying to cast that to `List<Element>`. Element is a Node.

This change in dom4j is the only place I was able to find any breaking change. The ec2-plugin has a similar usage in production code. In that case, I did the generics type checking at runtime. For this test, I figured just let the test fail with ClassException if somehow the type isn't an Element.